pkg/plugin/datasource.go

```diff
@@ -0,0 +1,326 @@
+package plugin
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/criblcloud/search-datasource/pkg/models"
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana-plugin-sdk-go/backend/instancemgmt"
+	"github.com/grafana/grafana-plugin-sdk-go/backend/resource/httpadapter"
+	"github.com/grafana/grafana-plugin-sdk-go/data"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+// Ensure it actually implements the interfaces we need it to
+var (
+	_ backend.QueryDataHandler    = (*Datasource)(nil)
+	_ backend.CheckHealthHandler  = (*Datasource)(nil)
+	_ backend.CallResourceHandler = (*Datasource)(nil)
+)
+
+const MAX_RESULTS = 10000 // same as what the actual Cribl UI imposes
+const QUERY_PAGE_SIZE = 1000
+const CRIBL_TIME_FIELD = "_time"
+const MAX_DELAY_WAITING_FOR_FINISHED = 60 * time.Second // could make this configurable in a future release
+const MAX_BACKOFF_DURATION = 2 * time.Second
+const GRAFANA_TIME_FIELD_NAME = "Time"
+
+// Expose a counter metric tracking the # of queries, broken down by type (adhoc vs. savedSearchId)
+var queryCounter = promauto.NewCounterVec(
+	prometheus.CounterOpts{
+		Namespace: "grafana_plugin", // recommended by Grafana in their "Best Practices" section
+		Name:      "cribl_search_queries_total",
+		Help:      "Total number of queries.",
+	},
+	[]string{"query_type"},
+)
+
+// Expose a counter metric tracking the total # of result events produced by queries
+var resultsCounter = promauto.NewCounterVec(
+	prometheus.CounterOpts{
+		Namespace: "grafana_plugin",
+		Name:      "cribl_search_results_total",
+		Help:      "Total number of events in results.",
+	},
+	[]string{"query_type"},
+)
+
```

Comment:
NIT: DataSource (to match others)

Reply:
Which others?  This is straight out of Grafana codegen.